dpcpp/components/cooperative_groups.dp.hpp

```diff
@@ -0,0 +1,513 @@
+tiled_partition(const Group &group)
+{
+    return detail::thread_block_tile<Size>(group);
+}
+
+
+namespace detail {
+
+
+template <unsigned Size>
+struct is_group_impl<thread_block_tile<Size>> : std::true_type {};
+
+
+template <unsigned Size>
+struct is_synchronizable_group_impl<thread_block_tile<Size>> : std::true_type {
+};
+
+
+template <unsigned Size>
+struct is_communicator_group_impl<thread_block_tile<Size>> : std::true_type {};
+
+
+}  // namespace detail
+
+
+class thread_block {
+    friend __dpct_inline__ thread_block this_thread_block(sycl::nd_item<3> &);
+
+public:
+    __dpct_inline__ unsigned thread_rank() const noexcept { return data_.rank; }
+
+    __dpct_inline__ unsigned size() const noexcept { return data_.size; }
+
+    __dpct_inline__ void sync() const noexcept { group_.barrier(); }
+
+private:
+    __dpct_inline__ thread_block(sycl::nd_item<3> &group)
+        : group_{group},
+          data_{static_cast<unsigned>(group.get_local_range().size()),
+                static_cast<unsigned>(group.get_local_linear_id())}
+    {}
+    struct alignas(8) {
+        unsigned size;
+        unsigned rank;
+    } data_;
+
+    sycl::nd_item<3> &group_;
+};
+
+
+__dpct_inline__ thread_block this_thread_block(sycl::nd_item<3> &group)
+{
+    return thread_block(group);
+}
+
+
+namespace detail {
+
+
+template <>
+struct is_group_impl<thread_block> : std::true_type {};
+
+
+template <>
+struct is_synchronizable_group_impl<thread_block> : std::true_type {};
+
+
+}  // namespace detail
+
+
+/**
+ * This is a limited implementation of the DPCPP grid_group that works even on
+ * devices that do not support device-wide synchronization and without special
+ * kernel launch syntax.
+ *
+ * Note that this implementation does not support large grids, since it uses 32
+ * bits to represent sizes and ranks, while at least 73 bits (63 bit grid + 10
+ * bit block) would have to be used to represent the full space of thread ranks.
+ */
+class grid_group {
+    friend __dpct_inline__ grid_group this_grid(sycl::nd_item<3> &);
+
+public:
+    __dpct_inline__ unsigned size() const noexcept { return data_.size; }
+
+    __dpct_inline__ unsigned thread_rank() const noexcept { return data_.rank; }
+
+private:
+    __dpct_inline__ grid_group(sycl::nd_item<3> &group)
+        : data_{static_cast<unsigned>(group.get_global_range().size()),
+                static_cast<unsigned>(group.get_global_linear_id())}
+    {}
+
+    struct alignas(8) {
+        unsigned size;
+        unsigned rank;
+    } data_;
+};
+
+// Not using this, as grid_group is not universally supported.
+// grid_group this_grid()
+// using cooperative_groups::this_grid;
+// Instead, use our limited implementation:
+__dpct_inline__ grid_group this_grid(sycl::nd_item<3> &group)
+{
+    return grid_group(group);
+}
+
+
+}  // namespace group
+}  // namespace dpcpp
+}  // namespace kernels
+}  // namespace gko
+
+
+// Enable group can directly use group function
+__SYCL_INLINE_NAMESPACE(cl)
```

Comment:
for example,  we use the group algorithm the reduce in ballot.
If we do not do this way, when we need use oneAPI implemenation on gko's subgroup, we need to add static_cast to subgroup type.
I agree with you, they might change it at some point.

Reply:
I stay this version first because it still works and it makes ginkgo cooperative group like an extension not another object.
But if they change frequently and we are hard to stay with them, we can delete them and use static_cast.